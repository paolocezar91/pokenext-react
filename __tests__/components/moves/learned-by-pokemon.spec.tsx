import LearnedByPokemon from '@/components/moves/learned-by-pokemon';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { INamedApiResource, IPokemon } from 'pokeapi-typescript';
import { mockAllIsIntersecting } from "react-intersection-observer/test-utils";

jest.mock("next-intl", () => ({
  useTranslations: () => (key: string) => key,
  useLocale: () => 'en'
}));

jest.mock('@/context/user-context', () => ({
  useUser: () => ({
    settings: { descriptionLang: 'en', artworkUrl: 'official-artwork', typeArtworkUrl: 'sword-shield' }
  })
}));

jest.mock('@tanstack/react-query', () => ({
  useQuery: () => ({
    data: [
      { name: "ditto", id: 132, types: [ { type: { name: "normal", url: "/api/v2/type/1/" } } ] },
      { name: "mew", id: 151, types: [ { type: { name: "psychic",  url: "/api/v2/type/14/" } } ] }
    ]
  })
}));

const mockLearnedByPokemon = [
    {
      name: "ditto",
      url: "/api/v2/pokemon/132/"
    },
    {
      name: "mew",
      url: "/api/v2/pokemon/151/"
    }
  ] as unknown as INamedApiResource<IPokemon>[];

describe('LearnedByPokemon Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the component with initial state', () => {
    render(<LearnedByPokemon pokemonList={mockLearnedByPokemon} />);
    mockAllIsIntersecting(true);

    // Test title transformation
    expect(screen.getByRole('heading')).toHaveTextContent('moves.learnedBy.title');
    expect(screen.getByText('Mew')).toBeInTheDocument();
    expect(screen.getByText('Ditto')).toBeInTheDocument();
  });

  it('should have the link to the pokemons', () => {
    render(<LearnedByPokemon pokemonList={mockLearnedByPokemon} />);
    mockAllIsIntersecting(true);

    // Test title transformation
    expect(screen.getByRole('heading')).toHaveTextContent('moves.learnedBy.title');
    expect(screen.getByText('Mew')).toBeInTheDocument();
    const mewLink = screen.getByRole('link', { name: 'Mew' });
    expect(mewLink).toHaveAttribute('href', '/en/pokedex/mew');
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<LearnedByPokemon pokemonList={mockLearnedByPokemon} />);
    mockAllIsIntersecting(true);
    expect(asFragment()).toMatchSnapshot();
  });
});