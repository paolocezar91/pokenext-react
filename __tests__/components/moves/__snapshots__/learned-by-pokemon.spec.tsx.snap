// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LearnedByPokemon Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-fit learned-by-pokemon w-full flex flex-col flex-1 h-0 mt-2"
  >
    <h3
      class="w-fit text-lg mb-4"
    >
      moves.learnedBy.title
    </h3>
    <table
      class="w-full text-xs "
    >
      <thead>
        <tr
          class="bg-black pt-1 sticky top-0 bg-background z-1"
        >
          <th
            class="bg-(--pokedex-red-dark) w-[5%]"
          />
          <th
            class="bg-(--pokedex-red-dark) w-[1%] text-white text-center px-2 py-1"
          >
            <div
              class="tooltip-container"
            >
              <div
                class="tooltip-trigger"
              >
                <button
                  class="
          flex 
          items-start
          ml-1
          px-2
          py-1
          cursor-pointer
          rounded
          bg-(--pokedex-red-dark)
          hover:bg-(--pokedex-red-darker)
          transition-colors
          
         btn btn-primary"
                  type="button"
                >
                  <span
                    class="mr-1"
                  >
                    #
                  </span>
                  <svg
                    aria-hidden="true"
                    class="w-5"
                    data-slot="icon"
                    fill="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M6.97 2.47a.75.75 0 0 1 1.06 0l4.5 4.5a.75.75 0 0 1-1.06 1.06L8.25 4.81V16.5a.75.75 0 0 1-1.5 0V4.81L3.53 8.03a.75.75 0 0 1-1.06-1.06l4.5-4.5Zm9.53 4.28a.75.75 0 0 1 .75.75v11.69l3.22-3.22a.75.75 0 1 1 1.06 1.06l-4.5 4.5a.75.75 0 0 1-1.06 0l-4.5-4.5a.75.75 0 1 1 1.06-1.06l3.22 3.22V7.5a.75.75 0 0 1 .75-.75Z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </th>
          <th
            class="bg-(--pokedex-red-dark) w-[15%] text-white text-left px-2 py-1"
          >
            <div
              class="tooltip-container"
            >
              <div
                class="tooltip-trigger"
              >
                <button
                  class="
          flex 
          items-start
          ml-1
          px-2
          py-1
          cursor-pointer
          rounded
          bg-(--pokedex-red-dark)
          hover:bg-(--pokedex-red-darker)
          transition-colors
          
         btn btn-primary"
                  type="button"
                >
                  <span
                    class="mr-1"
                  >
                    table.name
                  </span>
                  <svg
                    aria-hidden="true"
                    class="w-5"
                    data-slot="icon"
                    fill="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M6.97 2.47a.75.75 0 0 1 1.06 0l4.5 4.5a.75.75 0 0 1-1.06 1.06L8.25 4.81V16.5a.75.75 0 0 1-1.5 0V4.81L3.53 8.03a.75.75 0 0 1-1.06-1.06l4.5-4.5Zm9.53 4.28a.75.75 0 0 1 .75.75v11.69l3.22-3.22a.75.75 0 1 1 1.06 1.06l-4.5 4.5a.75.75 0 0 1-1.06 0l-4.5-4.5a.75.75 0 1 1 1.06-1.06l3.22 3.22V7.5a.75.75 0 0 1 .75-.75Z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </th>
          <th
            class="bg-(--pokedex-red-dark) w-[10%] text-white text-left px-2 py-1"
          >
            <div
              class="tooltip-container"
            >
              <div
                class="tooltip-trigger"
              >
                <button
                  class="
          flex 
          items-start
          ml-1
          px-2
          py-1
          cursor-pointer
          rounded
          bg-(--pokedex-red-dark)
          hover:bg-(--pokedex-red-darker)
          transition-colors
          
         btn btn-primary"
                  type="button"
                >
                  <span
                    class="mr-1"
                  >
                    table.types
                  </span>
                  <svg
                    aria-hidden="true"
                    class="w-5"
                    data-slot="icon"
                    fill="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M6.97 2.47a.75.75 0 0 1 1.06 0l4.5 4.5a.75.75 0 0 1-1.06 1.06L8.25 4.81V16.5a.75.75 0 0 1-1.5 0V4.81L3.53 8.03a.75.75 0 0 1-1.06-1.06l4.5-4.5Zm9.53 4.28a.75.75 0 0 1 .75.75v11.69l3.22-3.22a.75.75 0 1 1 1.06 1.06l-4.5 4.5a.75.75 0 0 1-1.06 0l-4.5-4.5a.75.75 0 1 1 1.06-1.06l3.22 3.22V7.5a.75.75 0 0 1 .75-.75Z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="border-solid border-foreground  border-b-2"
        >
          <td
            class="p-2"
          >
            <a
              href="/pokedex/ditto"
            >
              <div
                class="
        pokemon flex flex-col mx-auto justify-center items-center
        
        w-30 h-30
        xs
        titleless
      "
                style="background: rgb(168, 167, 122);"
              >
                <div
                  class="pokemon-shadow bg-[rgba(0,0,0,0.2)] w-30 h-30"
                />
                <div
                  class="img-hover-zoom w-full h-full"
                >
                  <div
                    class="relative w-full h-[120px]"
                  >
                    <img
                      alt="Ditto #0132"
                      class="artwork z-1 official-artwork"
                      data-nimg="fill"
                      decoding="async"
                      loading="lazy"
                      sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
                      src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=3840&q=75"
                      srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F132.png&w=3840&q=75 3840w"
                      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                    />
                  </div>
                </div>
              </div>
            </a>
          </td>
          <td
            class="p-2 text-center"
          >
            #0132
          </td>
          <td
            class="p-2"
          >
            <a
              class="transition-colors hover:bg-(--pokedex-red-dark) p-1"
              href="/pokedex/ditto"
            >
              Ditto
            </a>
          </td>
          <td
            class="p-2"
          >
            <a
              href="/type/normal"
            >
              <img
                alt="Normal"
                class="inline m-1"
                data-nimg="1"
                decoding="async"
                height="20"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Ftypes%2Fgeneration-viii%2Fsword-shield%2F1.png&w=256&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Ftypes%2Fgeneration-viii%2Fsword-shield%2F1.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Ftypes%2Fgeneration-viii%2Fsword-shield%2F1.png&w=256&q=75 2x"
                style="color: transparent;"
                width="100"
              />
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="p-2"
          >
            <a
              href="/pokedex/mew"
            >
              <div
                class="
        pokemon flex flex-col mx-auto justify-center items-center
        
        w-30 h-30
        xs
        titleless
      "
                style="background: rgb(249, 85, 135);"
              >
                <div
                  class="pokemon-shadow bg-[rgba(0,0,0,0.2)] w-30 h-30"
                />
                <div
                  class="img-hover-zoom w-full h-full"
                >
                  <div
                    class="relative w-full h-[120px]"
                  >
                    <img
                      alt="Mew #0151"
                      class="artwork z-1 official-artwork"
                      data-nimg="fill"
                      decoding="async"
                      loading="lazy"
                      sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
                      src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=3840&q=75"
                      srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2Fother%2Fofficial-artwork%2F151.png&w=3840&q=75 3840w"
                      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                    />
                  </div>
                </div>
              </div>
            </a>
          </td>
          <td
            class="p-2 text-center"
          >
            #0151
          </td>
          <td
            class="p-2"
          >
            <a
              class="transition-colors hover:bg-(--pokedex-red-dark) p-1"
              href="/pokedex/mew"
            >
              Mew
            </a>
          </td>
          <td
            class="p-2"
          >
            <a
              href="/type/psychic"
            >
              <img
                alt="Psychic"
                class="inline m-1"
                data-nimg="1"
                decoding="async"
                height="20"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Ftypes%2Fgeneration-viii%2Fsword-shield%2F14.png&w=256&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Ftypes%2Fgeneration-viii%2Fsword-shield%2F14.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Ftypes%2Fgeneration-viii%2Fsword-shield%2F14.png&w=256&q=75 2x"
                style="color: transparent;"
                width="100"
              />
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
